snuba/snuba_migrations/events_analytics_platform/0011_span_attribute_table.py

```diff
@@ -0,0 +1,126 @@
+    The table is populated by a separate materialized view for each type of attribute.
+    """
+
+    blocking = False
+    storage_set_key = StorageSetKey.EVENTS_ANALYTICS_PLATFORM
+    granularity = "8192"
+
+    meta_view_name = "spans_str_attrs_mv"
+    meta_local_table_name = "spans_str_attrs_local"
+    meta_dist_table_name = "spans_str_attrs_dist"
+    meta_table_columns: Sequence[Column[Modifiers]] = [
+        Column("organization_id", UInt(64)),
+        Column("project_id", UInt(64)),
+        Column(
+            "trace_id", UUID()
+        ),  # recommended by altinity, this lets us find traces which have k=v set
+        Column("attr_key", String()),
+        Column("attr_value", String()),
+        Column("timestamp", DateTime(modifiers=Modifiers(codecs=["ZSTD(1)"]))),
+        Column("retention_days", UInt(16)),
+        Column("count", SimpleAggregateFunction("sum", [UInt(64)])),
+    ]
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=self.storage_set_key,
+                table_name=self.meta_local_table_name,
+                engine=table_engines.AggregatingMergeTree(
+                    storage_set=self.storage_set_key,
+                    primary_key="(organization_id, attr_key)",
+                    order_by="(organization_id, attr_key, attr_value, timestamp, trace_id, project_id, retention_days)",
+                    partition_by="toMonday(timestamp)",
+                    settings={
+                        "index_granularity": self.granularity,
+                    },
+                    ttl="timestamp + toIntervalDay(retention_days)",
+                ),
+                columns=self.meta_table_columns,
+                target=OperationTarget.LOCAL,
+            ),
+            operations.CreateTable(
+                storage_set=self.storage_set_key,
+                table_name=self.meta_dist_table_name,
+                engine=table_engines.Distributed(
+                    local_table_name=self.meta_local_table_name, sharding_key=None
+                ),
+                columns=self.meta_table_columns,
+                target=OperationTarget.DISTRIBUTED,
+            ),
+            operations.CreateMaterializedView(
+                storage_set=self.storage_set_key,
+                view_name=self.meta_view_name,
+                columns=self.meta_table_columns,
+                destination_table_name=self.meta_local_table_name,
+                target=OperationTarget.LOCAL,
+                query=f"""
+SELECT
+    organization_id,
+    project_id,
+    trace_id,
+    attrs.1 as attr_key,
+    attrs.2 as attr_value,
+    toStartOfDay(_sort_timestamp) AS timestamp,
+    retention_days,
+    1 AS count
+FROM eap_spans_local
+LEFT ARRAY JOIN
+    arrayConcat({",".join(f"CAST(attr_str_{n}, 'Array(Tuple(String, String))')" for n in range(ATTRIBUTE_BUCKETS))}) AS attrs
```

Comment:
What is the thinking behind arrayConcat vs how Evan did it?

https://github.com/getsentry/snuba/pull/6152/files#diff-8174280b297958e24d650d9555da2a5026239db01793cf1ab6be4146ef504bbdR94-R99

Reply:
Evan still did arrayconcat ([L22 of that file](https://github.com/getsentry/snuba/pull/6152/files#diff-8174280b297958e24d650d9555da2a5026239db01793cf1ab6be4146ef504bbdR22)), this just does it with one less copy since the tuples can be shared as-is, instead of being split and re-joined into a list